SIGSEGV with azurerm_storage_account creation when empty share_properties block is provided.

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

<!--- Please keep this note for the community --->

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->


### Terraform Version

1.3.4

### AzureRM Provider Version

3.30.0

### Affected Resource(s)/Data Source(s)

azurerm_storage_account

### Terraform Configuration Files

```hcl
terraform {
  required_version = ">= 1.3.0"

  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~> 3.0"
    }
  }
}

provider "azurerm" {
  features {}
}

resource "azurerm_storage_account" "myaccount" {
  resource_group_name      = var.rgName
  name                     = var.name
  location                 = "westus2"
  account_tier             = "Standard"
  account_replication_type = "LRS"
  account_kind             = "StorageV2"

  share_properties {
  }
}

variable "rgName" {
  default = "awang"
} 

variable "name" {
  default = "awangtestshare"
}
```


### Debug Output/Panic Output

```shell
azurerm_storage_account.myaccount: Creating...
azurerm_storage_account.myaccount: Still creating... [10s elapsed]
azurerm_storage_account.myaccount: Still creating... [20s elapsed]
azurerm_storage_account.myaccount: Still creating... [30s elapsed]
╷
│ Error: Request cancelled
│ 
│   with azurerm_storage_account.myaccount,
│   on main.tf line 16, in resource "azurerm_storage_account" "myaccount":
│   16: resource "azurerm_storage_account" "myaccount" {
│ 
│ The plugin.(*GRPCProvider).ApplyResourceChange request was cancelled.
╵

Stack trace from the terraform-provider-azurerm_v3.30.0_x5 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x106b7428c]

goroutine 136 [running]:
github.com/hashicorp/terraform-provider-azurerm/internal/services/storage.resourceStorageAccountCreate(0x140015b8f78?, {0x107944ca0?, 0x14000754800})
	github.com/hashicorp/terraform-provider-azurerm/internal/services/storage/storage_account_resource.go:1317 +0x306c
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x108381240?, {0x108381240?, 0x14000be1b90?}, 0xd?, {0x107944ca0?, 0x14000754800?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.18.0/helper/schema/resource.go:695 +0x138
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0x14001679960, {0x108381240, 0x14000be1b90}, 0x1400194b790, 0x14000ebdd00, {0x107944ca0, 0x14000754800})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.18.0/helper/schema/resource.go:837 +0x874
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0x14000788900, {0x108381240?, 0x14000be1ad0?}, 0x1400224b2c0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.18.0/helper/schema/grpc_provider.go:1021 +0xb94
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0x14000ab7860, {0x108381240?, 0x14000be1380?}, 0x14001948150)
	github.com/hashicorp/terraform-plugin-go@v0.10.0/tfprotov5/tf5server/server.go:813 +0x370
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x107f8ef20?, 0x14000ab7860}, {0x108381240, 0x14000be1380}, 0x140015ae960, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.10.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x174
google.golang.org/grpc.(*Server).processUnaryRPC(0x14000138a80, {0x1083934d8, 0x14000bd4680}, 0x140019386c0, 0x140017d8120, 0x10c4e7400, 0x0)
	google.golang.org/grpc@v1.47.0/server.go:1283 +0xb64
google.golang.org/grpc.(*Server).handleStream(0x14000138a80, {0x1083934d8, 0x14000bd4680}, 0x140019386c0, 0x0)
	google.golang.org/grpc@v1.47.0/server.go:1620 +0x840
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.47.0/server.go:922 +0x88
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.47.0/server.go:920 +0x298

Error: The terraform-provider-azurerm_v3.30.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```


### Expected Behaviour

shoudl work or error out gracefully

### Actual Behaviour

SIGSEGV

### Steps to Reproduce

use terraform apply against the provided main.tf.

The culprit is this block
```
  share_properties {
  }
```

If i remove it it's fine.  That's a reasonable solution but I wanted to report  this bug as a SIGSEGV crash is not ideal error handling.

### Important Factoids

_No response_

### References

_No response_

--- a/internal/services/storage/storage_account_resource.go
+++ b/internal/services/storage/storage_account_resource.go
@@ -1336,17 +1336,22 @@ func resourceStorageAccountCreate(d *pluginsdk.ResourceData, meta interface{}) e
 		fileServiceClient := meta.(*clients.Client).Storage.FileServicesClient
 
 		shareProperties := expandShareProperties(val.([]interface{}))
+
 		// The API complains if any multichannel info is sent on non premium fileshares. Even if multichannel is set to false
-		if accountTier != string(storage.SkuTierPremium) {
+		if accountTier != string(storage.SkuTierPremium) && shareProperties.FileServicePropertiesProperties.ProtocolSettings != nil {
 
 			// Error if the user has tried to enable multichannel on a standard tier storage account
-			if shareProperties.FileServicePropertiesProperties.ProtocolSettings.Smb.Multichannel != nil && shareProperties.FileServicePropertiesProperties.ProtocolSettings.Smb.Multichannel.Enabled != nil {
-				if *shareProperties.FileServicePropertiesProperties.ProtocolSettings.Smb.Multichannel.Enabled {
-					return fmt.Errorf("`multichannel_enabled` isn't supported for Standard tier Storage accounts")
-				}
-			}
+			smb := shareProperties.FileServicePropertiesProperties.ProtocolSettings.Smb
+			if smb != nil && smb.Multichannel != nil {
 
-			shareProperties.FileServicePropertiesProperties.ProtocolSettings.Smb.Multichannel = nil
+				if smb.Multichannel.Enabled != nil {
+					if *shareProperties.FileServicePropertiesProperties.ProtocolSettings.Smb.Multichannel.Enabled {
+						return fmt.Errorf("`multichannel_enabled` isn't supported for Standard tier Storage accounts")
+					}
+				}
+
+				shareProperties.FileServicePropertiesProperties.ProtocolSettings.Smb.Multichannel = nil
+			}
 		}
 
 		if _, err = fileServiceClient.SetServiceProperties(ctx, id.ResourceGroup, id.Name, shareProperties); err != nil {

--- a/internal/services/storage/storage_account_resource.go
+++ b/internal/services/storage/storage_account_resource.go
@@ -647,6 +647,7 @@ func resourceStorageAccount() *pluginsdk.Resource {
 			"share_properties": {
 				Type:     pluginsdk.TypeList,
 				Optional: true,
+				// (@jackofallops) TODO - This should not be computed, however, this would be a breaking change with unknown implications for user data so needs to be addressed for 4.0
 				Computed: true,
 				MaxItems: 1,
 				Elem: &pluginsdk.Resource{
@@ -1338,7 +1339,7 @@ func resourceStorageAccountCreate(d *pluginsdk.ResourceData, meta interface{}) e
 		shareProperties := expandShareProperties(val.([]interface{}))
 
 		// The API complains if any multichannel info is sent on non premium fileshares. Even if multichannel is set to false
-		if accountTier != string(storage.SkuTierPremium) && shareProperties.FileServicePropertiesProperties.ProtocolSettings != nil {
+		if accountTier != string(storage.SkuTierPremium) && shareProperties.FileServicePropertiesProperties != nil && shareProperties.FileServicePropertiesProperties.ProtocolSettings != nil {
 
 			// Error if the user has tried to enable multichannel on a standard tier storage account
 			smb := shareProperties.FileServicePropertiesProperties.ProtocolSettings.Smb

--- a/internal/services/storage/storage_account_resource_test.go
+++ b/internal/services/storage/storage_account_resource_test.go
@@ -1350,13 +1350,13 @@ func TestAccStorageAccount_sasPolicy(t *testing.T) {
 	})
 }
 
-func TestAccStorageAccount_emptyStorageProperties(t *testing.T) {
+func TestAccStorageAccount_emptyShareProperties(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_storage_account", "test")
 	r := StorageAccountResource{}
 
 	data.ResourceTest(t, r, []acceptance.TestStep{
 		{
-			Config: r.emptyStorageProperties(data),
+			Config: r.emptyShareProperties(data),
 			Check: acceptance.ComposeTestCheckFunc(
 				check.That(data.ResourceName).ExistsInAzure(r),
 			),
@@ -4123,7 +4123,7 @@ resource "azurerm_storage_account" "test" {
 `, data.RandomInteger, data.Locations.Primary, data.RandomString)
 }
 
-func (r StorageAccountResource) emptyStorageProperties(data acceptance.TestData) string {
+func (r StorageAccountResource) emptyShareProperties(data acceptance.TestData) string {
 	return fmt.Sprintf(`
 provider "azurerm" {
   features {}

--- a/internal/services/storage/storage_account_resource_test.go
+++ b/internal/services/storage/storage_account_resource_test.go
@@ -1350,6 +1350,21 @@ func TestAccStorageAccount_sasPolicy(t *testing.T) {
 	})
 }
 
+func TestAccStorageAccount_emptyStorageProperties(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_storage_account", "test")
+	r := StorageAccountResource{}
+
+	data.ResourceTest(t, r, []acceptance.TestStep{
+		{
+			Config: r.emptyStorageProperties(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+	})
+}
+
 func (r StorageAccountResource) Exists(ctx context.Context, client *clients.Client, state *pluginsdk.InstanceState) (*bool, error) {
 	id, err := parse.StorageAccountID(state.ID)
 	if err != nil {
@@ -4104,6 +4119,31 @@ resource "azurerm_storage_account" "test" {
     expiration_action = "Log"
     expiration_period = "1.15:5:05"
   }
+}
+`, data.RandomInteger, data.Locations.Primary, data.RandomString)
+}
+
+func (r StorageAccountResource) emptyStorageProperties(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-storage-%[1]d"
+  location = "%[2]s"
+}
+
+resource "azurerm_storage_account" "test" {
+  name                = "unlikely23exst2acct%[3]s"
+  resource_group_name = azurerm_resource_group.test.name
+
+  location                 = azurerm_resource_group.test.location
+  account_tier             = "Standard"
+  account_replication_type = "LRS"
+  account_kind             = "StorageV2"
+
+  share_properties {}
 }
 `, data.RandomInteger, data.Locations.Primary, data.RandomString)
 }

